### ISSUE TITLE
Copy value of trainable variable to another trainable variable.

### ISSUE THREAD

**Issue by Firyuza** [tensorflow/tensorflow#23906]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 1.4.1
- Are you willing to contribute it (Yes/No): No.

**Describe the feature and the current behavior/state.**
Now Tensorflow doesn't allow to copy value of trainable variable to another one. Operations like tf.assight, tf.identity, tf.Variable(source_variable.initialized_value()) or tf.contrib.copy_graph.copy_variable_to_graph(var, self.graph) DO NOT copy just value. After such operations the destination variable referes to the source variable and I can't pass this variable to optimizer for computing gradients: "ValueError: No gradients provided for any variable, check your graph for ops that do not support gradients, ... "
**Will this change the current api? How?**
This change won't change the existing functions, just will add new opportunity for trainable variables.
**Who will benefit with this feature?**
It will make more flexible structure that can be good for GANs, or when we have two networks and it is neccessary to copy values of one network parameters to second network parameters, and still train two networks independently.
**Any Other info.**

**Comment by ebrevdo** [comment]
Please provide a **small** reproducible snippet of the problem.

**Comment by Firyuza** [comment]
Hi @ebrevdo !

I started working on several GANs architecture that based on [cvpr paper 2018](http://openaccess.thecvf.com/content_cvpr_2018/papers/Chavdarova_SGAN_An_Alternative_CVPR_2018_paper.pdf). 
1. I have _Discriminator_A_ and _Generator_A_. I train them.
2. Then I have to copy value of trainable variables of _Discriminator_A_ to trainable variable of _Discriminator_B_. Then I train _Discriminator_B_ and _Generator_B_.
In Tensorflow If I use **tf.assign**, **tf.Varaible** or even use two different **graphs** (tf.contrib.copy_graph.copy_variable_to_graph(var, self.graph)) I see that when I train _Discriminator_B_ and _Generator_B_ the _Discriminator_A_ is also training. So what I’ve got: actually _Discriminator_B_ refers to _Discriminator_A_ parameters, but I just need after every epoch take the values from _Discriminator_A_ and initialize _Discriminator_B_.

Example of code:
```
# original network from which I take values of trainable variables
D_vars = [var for var in self.train_vars if 'Discriminator_A' in var.name]

# The destination trainable variables
msg_D_vars = [var for var in self.train_vars if 'Discriminator_B' in var.name]

for var_idx, var in enumerate(D_vars):
     1)  tf.assign(msg_D_vars[var_idx], var)  -- DOESNT' WORK
     2) var = tf.contrib.copy_graph.copy_variable_to_graph(var, self.graph) -- DOESN'T WORK
    3) msg_D_vars[var_idx] = tf.Variable(var.initialized_value(), trainable=True) -- DOESN'T WORK
```

Also I tried to remove trainable variables before training after 1) or 2) or 3) step and then add them again for training. But this tricky approach doesn't work too as I need.
```
with self.graph.as_default():
        trainable_collection = tf.get_collection_ref(tf.GraphKeys.TRAINABLE_VARIABLES)
        variables_to_remove = list()
        for vari in trainable_collection:
             # uses the attribute 'name' of the variable
               if 'discriminator_B' % in vari.name:
                     variables_to_remove.append(vari)
       for rem in variables_to_remove:
           trainable_collection.remove(rem)
```

Thanks!

**Comment by zlichen** [comment]
@Firyuza You need to initialize a temp for exchange.

**Comment by zlichen** [comment]
Hello @Firyuza ! You might need to initialize a temp for exchange.
```
def hello(scopeName='sc1'):
    with tf.variable_scope(scopeName):
        a1 = tf.get_variable(name='test_var1', initializer=0.)
        b1 = tf.get_variable(name='test_var2', initializer=0.)



sess = tf.Session()
g=hello('sc1')
g2=hello('new')
g3=hello('old')
sess.run(tf.global_variables_initializer())
# params=tf.trainable_variables(scope='sc1')
t_vars=tf.trainable_variables()
var_in_sc1= [var for var in t_vars if 'sc1' in var.name]
var_in_new= [var for var in t_vars if 'new' in var.name]
var_in_old= [var for var in t_vars if 'old' in var.name]
print('save 0')
for var_idx, var in enumerate(var_in_sc1):
    sess.run(tf.assign(var_in_old[var_idx], var))

print('add 1')
for var in var_in_sc1:
    sess.run(tf.assign_add(var,1.))

print('save 1')
for var_idx, var in enumerate(var_in_sc1):
    sess.run(tf.assign(var_in_new[var_idx], var))

print('add 1')
for var in var_in_sc1:
    sess.run(tf.assign_add(var,1.))
print(sess.run(var_in_sc1))

print('load old')
for var_idx, var in enumerate(var_in_old):
    sess.run(tf.assign(var_in_sc1[var_idx], var))
print(sess.run(var_in_sc1))


print('load new')
for var_idx, var in enumerate(var_in_new):
    sess.run(tf.assign(var_in_sc1[var_idx], var))
print(sess.run(var_in_sc1))
```
Hope that this demo can help you!

**Comment by Firyuza** [comment]
Hello @zlichen!

Thank you for feedback!

Yes, this demo works. But I have to notice that it's not so good and intuitive way of making such kind of copy. I hope [Tensorflow](https://github.com/tensorflow) community will consider more flexible and obvious way of doing it. Otherwise I need to write this code on PyTorch, because it has such ability.

Also, we have to keep in memory temp variables and it's not effective. E.g. if I'm working with ResNet it will be really expensive memory storing trainable variables values from 50 or 100 layers.

Thanks!

**Comment by zlichen** [comment]
Hello @Firyuza !

After testing the demo, I also found that is really time-consuming while copying the weight from a huge Neural Network. The demo below is a better and more effective way which is no need for a tmp.


```
import tensorflow as tf
with tf.Session() as sess:
    def hello(scopeName='sc1'):
        with tf.variable_scope(scopeName):
            a1 = tf.get_variable(name='test_var1', initializer=0.)
            b1 = tf.get_variable(name='test_var2', initializer=0.)

    g=hello('sc1')
    g2=hello('new')
    g3=hello('old')
    sess.run(tf.global_variables_initializer())

    t_vars=tf.trainable_variables()
    var_in_sc1= [var for var in t_vars if 'sc1' in var.name]
    var_in_new= [var for var in t_vars if 'new' in var.name]
    var_in_old= [var for var in t_vars if 'old' in var.name]
    
    print('save 0')
    for var_idx, var in enumerate(var_in_sc1):
        var_in_old[var_idx].load(var_in_sc1[var_idx].eval(),sess)
    print('add 1')
    for var in var_in_sc1:
        sess.run(tf.assign_add(var,1.))

    print('save 1')
    for var_idx, var in enumerate(var_in_sc1):
        var_in_new[var_idx].load(var_in_sc1[var_idx].eval(),sess)
    print('add 1')
    for var in var_in_sc1:
        sess.run(tf.assign_add(var,1.))
    print(sess.run(var_in_sc1))

    print('load old')
    for var_idx, var in enumerate(var_in_old):
        var_in_sc1[var_idx].load(var_in_old[var_idx].eval(),sess)
    print(sess.run(var_in_sc1))

    print('load new')
    for var_idx, var in enumerate(var_in_new):
        var_in_sc1[var_idx].load(var_in_new[var_idx].eval(),sess)
    print(sess.run(var_in_sc1))
```
Maybe still need a few lines of code. As far as I know, Keras (base on tensorflow) can also easily get or set the weights for the models or layers with only one line of code.

**Comment by Firyuza** [comment]
Hi @zlichen !
Thanks again for feedback! Your example is working, but now I noticed that If I create two networks in the same Graph and start to train first network I see that the second one is training too. They do not have sharable variables. I think maybe one of the my first approaches is working too, just the problem is on another plane...

**Comment by zlichen** [comment]
> Hi @zlichen !
> 
> Thanks again for feedback! Your example is working, but now I noticed that If I create two networks in the same Graph and start to train first network I see that the second one is training too. They do not have sharable variables. I think maybe one of the my first approaches is working too, just the problem is on another plane...
> 
> Thanks!

Hi! I guess you have to set different scope names for your two network in order to make distinction.

**Comment by zlichen** [comment]
> Hi @zlichen !
> 
> Thanks again for feedback! Your example is working, but now I noticed that If I create two networks in the same Graph and start to train first network I see that the second one is training too. They do not have sharable variables. I think maybe one of the my first approaches is working too, just the problem is on another plane...
> 
> Thanks!

Hi! I guess you have to set different scope names for your two networks in order to make distinction.

**Comment by Firyuza** [comment]
Hi @zlichen !

**Comment by Firyuza** [comment]
Hi @zlichen !

Yes, I gave different scope names for two networks, but it doesn't work. I think it's a Tensorflow paradigm, we cannot create two networks in the **same** graph when you want to train it **independently**. And  all copy operations that I discussed above do not make sense. 

So if we wanna make copy value of variable -- your solution is right. But if we wanna make a copy of variable's value of one network to varaible of another one, there is another solution:

1) for training two different  networks **independently**, define two **different** **graphs**;
2) copying values of varables of one network to variables of another one:
`for var_idx, var in enumerate(D_vars):
        msg_D_vars[var_idx].load(D_vars[var_idx].eval(session=self.sess), self.sess_msg)`

Thanks!
